### PR TITLE
[py-numpy, py-scipy] Enable MKL & ARMpl

### PR DESCRIPTION
**Pull request by stephenmsachs**
- Issue  https://github.com/numpy/numpy/issues/22011 has been resolved and backported for 1.23.4
- ARMpl and MKL when added to `site.cfg` in numpy and via pkg-config (in meson build) to scipy work. Tested on x86_64 (icc) and aarch64 (gcc).

Depends on https://github.com/spack/spack/pull/34978

### CODE REVIEW

**Review by adamjstewart — COMMENTED** [review]
I don't see anything wrong with this, will let @rgommers review in more detail. MKL will be tested by CI.

**Comment by rgommers** [comment]
The three CI failures are all the same ([example log](https://gitlab.spack.io/spack/spack/-/jobs/5185794), and seem related to this PR:
```
     48    Library npymath found: YES
     49    Library npyrandom found: YES
     50    Did not find CMake 'cmake'
     51    Found CMake: NO
     52    Run-time dependency mkl-dynamic-lp64-seq found: NO (tried pkgconfig)
     53    
  >> 54    scipy/meson.build:131:0: ERROR: Dependency "mkl-dynamic-lp64-seq" no
           t found, tried pkgconfig
```

File: var/spack/repos/builtin/packages/py-scipy/package.py [hunk 1]
```diff
@@ -200,14 +195,16 @@ def setup_build_environment(self, env):
...
             blas = "mkl-dynamic-lp64-seq"
         if spec["lapack"].name in ["intel-mkl", "intel-parallel-studio", "intel-oneapi-mkl"]:
             lapack = "mkl-dynamic-lp64-seq"
-
         if spec["blas"].name in ["blis", "amdblis"]:
             blas = "blis"
+        if blas == "armpl":
+            blas += "-dynamic-lp64-seq"
+        if lapack == "armpl":
+            lapack += "-dynamic-lp64-seq"
```
Comment (rgommers) [inline var/spack/repos/builtin/packages/py-scipy/package.py:207]: The docs at https://developer.arm.com/downloads/-/arm-performance-libraries/get-started-with-armpl-free-version/single-page say that the correct name is `armpl_lp64`. And the link line in [this doc pagefor ILP64](https://developer.arm.com/documentation/101004/2210/General-information/Arm-Performance-Libraries-variant-using-64-bit-integer--INTEGER-8--arguments?lang=en) says `-larmpl_ilp64`. 

I can't find anything about pkg-config file names, but I wanted to check: are you sure this is right? If you could show a directory listing with all pkg-config files as shipped in ArmPL, that would be quite helpful. Both for review here, and for adding support in Meson for ArmPL, so we don't need this patch anymore in the future.

Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-scipy/package.py:207]: I should have been more clear about the last sentence: "Depends on https://github.com/spack/spack/pull/34978". Please take a look for reference.

ArmPL ships pkg-config files under `armpl_22.1_gcc-11.2/pkgconfig` and they don't have the `.pc` suffix. I am in contact with ARM to change that in the future.

Comment (rgommers) [reply var/spack/repos/builtin/packages/py-scipy/package.py:207]: Yes, I saw that and am following along with that PR. But it doesn't tell me what the file names are. Can you show the output of `ls armpl_22.1_gcc-11.2/pkgconfig`?

Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-scipy/package.py:207]: Directory content (after applying https://github.com/spack/spack/pull/34978):
```
[ec2-user@ip-10-7-99-193 ~]$ ls $(find $(spack location -i armpl-gcc) -name pkgconfig)
armpl-dynamic-ilp64-omp             armpl-Fortran-static-ilp64-omp
armpl-dynamic-ilp64-omp.pc          armpl-Fortran-static-ilp64-omp.pc
armpl-dynamic-ilp64-seq             armpl-Fortran-static-ilp64-seq
armpl-dynamic-ilp64-seq.pc          armpl-Fortran-static-ilp64-seq.pc
armpl-dynamic-lp64-omp              armpl-Fortran-static-lp64-omp
armpl-dynamic-lp64-omp.pc           armpl-Fortran-static-lp64-omp.pc
armpl-dynamic-lp64-seq              armpl-Fortran-static-lp64-seq
armpl-dynamic-lp64-seq.pc           armpl-Fortran-static-lp64-seq.pc
armpl-Fortran-dynamic-ilp64-omp     armpl-static-ilp64-omp
armpl-Fortran-dynamic-ilp64-omp.pc  armpl-static-ilp64-omp.pc
armpl-Fortran-dynamic-ilp64-seq     armpl-static-ilp64-seq
armpl-Fortran-dynamic-ilp64-seq.pc  armpl-static-ilp64-seq.pc
armpl-Fortran-dynamic-lp64-omp      armpl-static-lp64-omp
armpl-Fortran-dynamic-lp64-omp.pc   armpl-static-lp64-omp.pc
armpl-Fortran-dynamic-lp64-seq      armpl-static-lp64-seq
armpl-Fortran-dynamic-lp64-seq.pc   armpl-static-lp64-seq.pc
```

Comment (rgommers) [reply var/spack/repos/builtin/packages/py-scipy/package.py:207]: Great, thank you! I'll plan to add support for those names in Meson (xref https://github.com/mesonbuild/meson/pull/10921).

**Comment by stephenmsachs** [comment]
Re-basing onto develop now that https://github.com/spack/spack/pull/34978 has been merged.

**Comment by stephenmsachs** [comment]
CI failed because `intel-mkl` does not expose pkg-config files as `intel-oneapi-mkl` does. The last commit converges the bhavior. Adding @rscohn2 to review.

**Review by rgommers — APPROVED** [review]
LGTM, thanks @stephenmsachs. Should be good to go once Gitlab CI is happy.

**Review by adamjstewart — APPROVED** [review]

File: var/spack/repos/builtin/packages/py-numpy/package.py [hunk 2]
```diff
@@ -150,9 +150,9 @@ class PyNumpy(PythonPackage):
     # NVHPC support added in https://github.com/numpy/numpy/pull/17344
     conflicts("%nvhpc", when="@:1.19")
 
-    # Newer versions will not build with Intel https://github.com/numpy/numpy/issues/22011
-    conflicts("%intel", when="@1.23.0:")
-    conflicts("%oneapi", when="@1.23.0:")
+    # See https://github.com/numpy/numpy/issues/22011
+    conflicts("%intel", when="@1.23.0:1.23.3")
+    conflicts("%oneapi", when="@1.23.0:1.23.3")
```
Comment (rscohn2) [inline var/spack/repos/builtin/packages/py-numpy/package.py:155]: This PR does not change anything to make oneapi work. Do you know what the problem was ? PKG_CONFIG_PATH is set in intel-oneapi-mkl/package.py, but the gitlab tests were failing and I could not track it down.

Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-numpy/package.py:155]: Yes, that has nothing to do with the rest of this PR. That particular issue was a C++ not finding a few heapsort function calls. numpy changed them to be explicitly exported in a separate header file, so it works with intel&oneapi now.

**Pull request merged**